README.md
# private-telegram-bot

  Selective access telegram bots builder

  With private-telegram-bot you can manage your own Telegram bot with a security layer that allow build a public, selective or private bot.
Add or remove functionality to your bot is really easy. Add/remove any script in your public or private folder and you’ll can to run from your mobile. Easy!
Tested in Linux and Raspberry Pi!

## How install

Using npm: npm install private-telegram-bot

## First use

Before use it you must create a bot in Telegram bot platform. It’s very easy:

	- From your Telegram in your mobile, talk with @botFather
	- Say him: /newbot and follow his instructions
	- After bot creation, @botFather’ll give you a token. Save it! Its very important!

More info here: https://core.telegram.org/bots
With token of your new bot, edit /config/bot.json and add your token value to ‘token’ property.
Also, add in the same file, your Telegram alias or username (without '@') in 'owner' property.

Everything is ready… run with: ./private-telegram-bot.js
Now you can launch commands from your Telegram bot, from your mobile. Send /help to see which commands you have ready to run.

## How add your own commands

Easy. You can add any script in public or admin scripts folder. The name of your script will be a command.

By default, public and admin script folders are:
	[project folder]/publicScripts
  [project folder]/adminScripts

Only you (the bot owner) can run the ‘admins scripts’. Only who you want can run ‘public scripts’ (check ‘adduser’ and ‘removeuser’ commands)
You can change public and admin folder path properties in bot configuration file (bot.json)
You can add and remove commands, users, change configuration properties… without restart any process. Private-telegram-bot will detect and rebuild it any required object to get change.

## How to configure it

Only bot token is required to use it but, if you want to know how to modify bot behaviour, you can do it editing simple json property files in config folder.
Change bot language (currently ‘en’ and ‘es’ values), create rotate file log files, set commands timeout… etc Open your config files and take a look!


## License

(The MIT License)

Copyright (c) 2015 Guillermo Peña &lt;guillermo.pena.cardano@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

HISTORY
=======

0.0.1 / 2015-07-15 - Initial release
0.0.5 / 2015-07-15 - 'About' added
0.0.9 / 2015-07-16 - 'User without alias' bug fixed


lib/bot.js
/* ___________________________________________________________________________

  Module: bot.js   Version: v0.0.3
  Repository: http://github.com/GuillermoPena/private-telegram-bot
  Author: Guillermo Peña (guillermo.pena.cardano@gmail.com)
  Last update: 16/07/2015

  Bot Manager: create bot, filter and response messages...

____________________________________________________________________________*/

function Bot(cfgPath, logger){

  // ___ Modules ___________________________________________________________

  var cm = require('jsonfile-config-manager')         // Config Manager object
  var commandManager = require('./commandmanager.js') // Command Manager object
  var telegramBotAPI = require('telegram-bot-api')    // Telegram Bot API object
  var contains = require("multiple-contains")         // Contains library
  var path = require('path')                          // Path library

  // __ Properties _________________________________________________________

  var tba     // Telegram Bot API object
  var cmm     // Command Manager
  var lastMsg // Last message id

  var languages = ['es', 'en']

  // __ Private Methods ____________________________________________________

  // Build Bot
  var buildBot = function(args) {

    // Removing listeners if we are REbuilding bot
    if (tba) tba.removeAllListeners('message')

    // Verifying Bot language
    var language = 'en'
    if (!contains(languages, cm.config.bot.defaultLanguage, 'exists'))
      logger.print('warn', cm.config.literals.en.unknownLanguageError)
    else
      language = cm.config.bot.defaultLanguage

    // Verifying Bot token and owner
    var literal = cm.config.literals[language]
    if (!cm.config.bot.token) {
      logger.print('error', literal.tokenNotFoundError)
      return
    }
    if (!cm.config.bot.owner) {
      logger.print('error',  literal.ownerNotFoundError)
      return
    }

    // Building bot and retrieving bot info
    tba = new telegramBotAPI(cm.config.bot)
    tba.getMe(function(err, data) {
      if (err) {
        logger.print('error', literal.wrongTokenError)
        logger.printError(err)
        process.exit(3)
      }
      else {
        tba.username = data.username
        tba.firstname = data.first_name
        logger.print('info', tba.firstname + " ... OK")

        // Receiving messages
        tba.on('message', function(message){
          filterMessages(message)
        })
        logger.print('info', tba.firstname + " listening ...")
        cmm = new commandManager(tba, logger, cfgPath)
      }
    })
  }

  // Send literal errors or messages in bot language
  var sendLiteral = function(message, response, language) {

    // Getting literal
    if (!language) language = cm.config.bot.defaultLanguage
    var literal = cm.config.literals[language][response]

    // Replacing variables
    var username = (!message.from.username) ? message.from.id : message.from.username
    literal = literal.replace(new RegExp('%username%'), username )
    literal = literal.replace(new RegExp('%botname%'), tba.firstname )

    // Sending response
    var params = { "chat_id": message.chat.id, "text": literal }
    var response = tba.sendMessage(params, logger.printBotError)
  }

  // Filtering Messages
  var filterMessages = function(message) {

    // Formating message and preparing variables
    var username = (!message.from.username) ? message.from.id : message.from.username
    var messageToLog = logger.prepareMessage(message)

    // Checking message
    var isFromOwner = (username == cm.config.bot.owner)
    var isFromUser = contains(cm.config.users, username, 'exists')
    var isCommand = (message.text && message.text.charAt(0) == '/' && message.text.length > 1)
    var activeUser = cmm.getActiveUserIdx(username)

    // Parsing command
    if (isCommand) {
      message.text = message.text.replace("/","")
      var slices = message.text.split(' ')
      var command = slices.shift()
      if (slices.length > 0) var args = slices.join(' ')
    }

    // Filter: No Authorized Messages
    if ( !isFromOwner && !isFromUser ) {
      if (cm.config.bot.responseNoAuthorizedMessages) {
        logger.print('info', messageToLog)
        sendLiteral(message, "notAuthorizedUserError")
      }
      return
    }

    // If command require cancel current operation
    if (command == "cancel") {
      logger.print('info', messageToLog)
      if (activeUser > -1) {
        cmm.removeActiveUser(username)
        sendLiteral(message, "operationCancelledError")
      } else {
        sendLiteral(message, "nothingToCancelError")
      }
      return
    }

    // Active user, has a command in progress...
    if (activeUser > -1) {
      cmm.sendToChild(username, message.text)
      logger.print('info', messageToLog)
      return
    }

    // Filter: No commands (First char must be '/')
    if (!isCommand) {
      logger.print('info', messageToLog)
      sendLiteral(message, "notCommandError")
      return
    }

    // Filter: Start command == Welcome message
    if (command == 'start') {
      logger.print('info', messageToLog)
      sendLiteral(message, "wellcomeMessage")
      return
    }

    // Filter: Unknown commands
    var isPublicCommand = cmm.isPublicCommand(command)
    var isAdminCommand = cmm.isAdminCommand(command)
    if (!isPublicCommand && !isAdminCommand) {
      logger.print('info', messageToLog)
      sendLiteral(message, "unknownCommandError")
      return
    }

    // Filter: No Authorized User
    if (isAdminCommand && !isFromOwner){
      logger.print('info', messageToLog)
      sendLiteral(message, "commandOnlyForOwnerError")
      return
    }

    // Running Known command
    lastMsg = message.message_id
    var scriptPath = (isPublicCommand) ? cmm.publicScriptsPath : cmm.adminScriptsPath
    if (scriptPath.charAt(0) == '.') scriptPath = path.resolve(__dirname, scriptPath)
    if (args) args = [args]
    cmm.run(message, path.join(scriptPath, command), args)
    logger.print('info', messageToLog)
    return
  }

  // __ Main _______________________________________________________________

  // Loading config
  cm.addFile(cfgPath + 'literals.json', null, true)
  cm.addFile(cfgPath + 'users.json', null, true)
  cm.addFile(cfgPath + 'bot.json', null, true, buildBot)
}

module.exports = Bot


lib/usermanager.js
/* ___________________________________________________________________________

  Module: usermanager.js   Version: v0.0.1
  Repository: http://github.com/GuillermoPena/private-telegram-bot
  Author: Guillermo Peña (guillermo.pena.cardano@gmail.com)
  Last update: 1/07/2015

  User managment module

____________________________________________________________________________*/

function userManager(args) {

  // ___ Modules ___________________________________________________________


  // __ Properties _________________________________________________________

  var cm = args.configManager // Config Manager object
  var logger = args.logger    // Logger (winston) object
  var bot = args.bot          // Telegram Bot object

  // __ Private Methods ____________________________________________________


  // __ Public Methods _____________________________________________________

  // Add user
  var addUser = function() {

  }

  // Remove user
  var removeUser = function() {

  }

  // __ Return _____________________________________________________________

  return { "addUser": addUser
         , "removeUser": removeUser
         , "next": next
         }
}

module.exports = userManager


private-telegram-bot.js
#!/usr/bin/node

/* ___________________________________________________________________________

  Module: private-telegram-bot.js   Version: v0.0.3
  Repository: http://github.com/GuillermoPena/private-telegram-bot
  Author: Guillermo Peña (guillermo.pena.cardano@gmail.com)
  Last update: 16/07/2015

  Selective access telegram bots builder

____________________________________________________________________________*/

// Logger
var easyLogger = require('easylogger')
var logger = new easyLogger(__dirname + '/config/logger.json')

// Message bot formatter
logger.prepareMessage = function(message, to) {

  var messageToLog = ""
  // If NOT error...
  if (message.message_id) {

    // If last char is \n, remove it
    if (message.text.charAt(message.text.length-1) == '\n')
      message.text = message.text.substring(0, message.text.length-1)

    // Adding header in every line
    var username = (!message.from.username) ? message.from.id : message.from.username
    var header = "[msg_" + message.message_id + "] @" + username
    header += (to) ? " > @" + to + " : " : " : "
    messageToLog += header
    if (message.text) {
      message.text = message.text.replace(/\n/g,'\n' + header)
      messageToLog += message.text
    }
  }

  // If error...
  if (message.code)
    messageToLog = message.description.replace("Error:", "Error " + message.code + ":")

  return messageToLog
}

// Message bot printer
logger.printBotError = function(err, data, callback) {

  // Preparing message to be log
  var to = (err) ? null : data.chat.username
  var message = (err) ? logger.prepareMessage(err, to)
                      : logger.prepareMessage(data, to)
  var level = (err) ? 'error' : 'info'

  // Logging...
  logger.print(level, message)
}


// Preparing to catch error events
process.on('SIGSTP', function()  {
  logger.print('warn', 'Stopping Bot due to SIGSTP signal')
  process.exit(1)
})
process.on('SIGINT', function()  {
  logger.print('warn', 'Stopping Bot due to SIGINT signal')
  process.exit(1)
})
process.on('SIGTERM', function() {
  logger.print('warn', 'Stopping Bot due to SIGTERM signal')
  process.exit(1)
})
process.on('SIGTSTP', function() {
  logger.print('warn', 'Stopping Bot due to SIGTSTP signal')
  process.exit(1)
})
/*
process.on('uncaughtException', function(e) {
  logger.print('error', 'Stopping Bot due to uncaughtException')
  logger.print('error', require('util').inspect(e))
  process.exit(128)
})
*/

// Launching bot...
var Bot = require(__dirname + '/lib/bot.js') // Bot Library
var bot = new Bot(__dirname + '/config/', logger)


lib/commandmanager.js
/* ___________________________________________________________________________

  Module: commandManager.js   Version: v0.0.3
  Repository: http://github.com/GuillermoPena/private-telegram-bot
  Author: Guillermo Peña (guillermo.pena.cardano@gmail.com)
  Last update: 16/07/2015

  Module to run commands, catch responses and send from Telegram Bot to user

  __________________________________________________________________________*/

function CommandManager(bot, logger, configPath) {

  // ___ Modules ___________________________________________________________

  var fs = require('fs')                      // File System library
  var cm = require('jsonfile-config-manager') // Config Manager object
  var chokidar = require('chokidar')          // Chokidar library (FS Watcher)
  var contains = require("multiple-contains") // Contains library
  var path = require('path')                  // Path library

  // __ Private Methods ____________________________________________________

  // Translate exit codes
  var checkExitCode = function(code) {
    // TODO: check exit code of child process in error and exit events
  }

  // Function to re-send output and error messages
  var sendToUser = function(data, message) {

    var params = { "chat_id": message.chat.id, "text": data }
    var response = bot.sendMessage(params, logger.printBotError)
  }

  // Is script file executable?
  var isExecutable = function(file) {

    // Extracting execution permissions for owner and other users
    var fstat = fs.statSync(file)
    var decimalPermissions = parseInt(fstat.mode.toString(8), 10).toString()
    var ownerExeChar = decimalPermissions.charAt(3)
    var otherExeChar = decimalPermissions.charAt(5)

    return ( ownerExeChar == '7' || ownerExeChar == '5' ||
             otherExeChar == '7' || otherExeChar == '5' )
  }

  // Add command
  var addCommand = function(publicOrAdmin, file) {

    if (isExecutable(file)) {
      // Getting script file name
      var command = file.split('/').pop()
      logger.print('debug', 'New ' + publicOrAdmin + ' command detected has been added : ' + file)
      if (!contains(cm.config.help.commands, {"key":"name", "value":command}, 'exists'))
        logger.print('warn', command + ' : Dont forget add detailed help in help.json file')

      // Adding to public/admin commands array
      var commands = cm.config.commands[publicOrAdmin + 'Commands']
      commands.push(command)
    }
  }

  // Remove command
  var removeCommand = function(publicOrAdmin, file, removeHelp) {

    // Getting script file name
    var script = file.split('/').pop()
    logger.print('debug', 'One ' + publicOrAdmin + ' command has been removed : ' + script)

    // Removing from public/admin commands array
    var commands = cm.config.commands[publicOrAdmin + 'Commands']
    var cid = contains(commands, script, 'index')
    if (cid > -1) commands.splice(cid, 1)

    // Removing info in help.json
    if (removeHelp) {
      var helpIdx = contains( cm.config.help.commands
                            , { "key": "name", "value": script }
                            , "index")
      if (helpIdx > -1) {
        cm.config.help.commands.splice(helpIdx, 1)
        cm.updateBranchFile('help')
      }
    }
  }

  // Add command in progress to array
  var addActiveUser = function(username, command, child, timeout) {

    // Adding active user
    activeUsers.push({ "username": username
                     , "command": command
                     , "child": child
                     , "timeout": timeout
                     })
  }

  // Build Command Manager
  var buildCommandManager = function() {

    cm.config.commands.adminCommands = []     // Array of Admin Commands
    cm.config.commands.publicCommands = []    // Array of Public Commands

    // Removing listeners if we are REbuilding command Manager
    if (publicWatcher) {
      publicWatcher.removeAllListeners('add')
      publicWatcher.removeAllListeners('unlink')
    }
    if (adminWatcher) {
      adminWatcher.removeAllListeners('add')
      adminWatcher.removeAllListeners('unlink')
    }

    // Watch public script folder
    var publicScriptsPath = (cm.config.commands.publicScriptsPath.charAt(0) == '.') ?
                            path.resolve(__dirname, cm.config.commands.publicScriptsPath)
                          : cm.config.commands.publicScriptsPath
    var publicWatcher = chokidar.watch( publicScriptsPath
                                      , {ignored: /^\./, persistent: true})
    publicWatcher
      .on('add',    function(file)  { addCommand('public', file) })
      .on('unlink', function(file)  { removeCommand('public', file, true) })
      .on('change', function(file)  {
        removeCommand('public', file, false)
        addCommand('public', file)
      })

    // Watch owner script folder
    var adminScriptsPath = (cm.config.commands.adminScriptsPath.charAt(0) == '.') ?
                            path.resolve(__dirname,  cm.config.commands.adminScriptsPath)
                          : cm.config.commands.adminScriptsPath
    var adminWatcher = chokidar.watch( adminScriptsPath
                                     , {ignored: /^\./, persistent: true})
    adminWatcher
      .on('add',    function(file)  { addCommand('admin', file) })
      .on('unlink', function(file)  { removeCommand('admin', file, true) })
      .on('change', function(file)  {
        removeCommand('admin', file, false)
        addCommand('admin', file)
      })
  }

  // __ Public Methods _____________________________________________________

  // Run script file
  var run = function(message, scriptFile, args) {

    // Creating child process
    scriptFile = path.resolve(__dirname, scriptFile)

    var username = (!message.from.username) ? message.from.id : message.from.username
    var command  = scriptFile.split('/').pop()
    logger.print('debug', "User @" + username + " is running command '" + command + ' ' + args + "'")
    var child = require('child_process').execFile(scriptFile, args, function(err,data){
      removeActiveUser(username)
    })

    // Timeout for command
    var timeout = setTimeout( function() {
      var literal = cm.config.literals[cm.config.bot.defaultLanguage]
      var params = { "chat_id": message.chat.id
                   , "text": '/' + command + " : " + literal.commandTimeoutError }
      var response = bot.sendMessage(params, logger.printBotError)
      removeActiveUser(username)
    }, cm.config.commands.commandTimeout )

    // Returning child data/error output to user
    child.stdout.on('data', function(data) { sendToUser(data, message) })
    child.stderr.on('data', function(data) { sendToUser(data, message) })

    child.once('error', function(err) {
      logger.print('error', 'Command ' + command + ' : Error : ' + err)
      sendToUser(err, message)
      removeActiveUser(username)
    })

    // Adding active user to array
    addActiveUser(username, command, child, timeout)
  }

  // Get index of an user running a command?
  var getActiveUserIdx = function(username) {
    return idx = contains( activeUsers
                         , { "key": "username", "value": username }
                         , "index"
                         )
  }

  // Remove command from array after stop it
  var removeActiveUser = function(username) {

    // Getting active user by username
    var idx = getActiveUserIdx(username)
    if (idx > -1) {

      // Clearing timeout
      clearTimeout(activeUsers[idx].timeout)

      // Removing events and killing child
      var child = activeUsers[idx].child
      logger.print('debug', 'Removing active user ' + username + ' from array. Index: ' + idx)

      // Removing listeners
      setTimeout(function() {
        child.removeAllListeners('data')
        child.removeAllListeners('error')

        // Check if child process is finished and kill it
        if (child.connected) child.kill('SIGINT')
      }, 300)

      // Removing active users from Array
      activeUsers.splice(idx, 1)
    }
  }

  // Send message to child proccess
  var sendToChild = function(username, message) {

    var idx = getActiveUserIdx(username)
    if (idx > -1 && message) {
      if (message.charAt(message.length-1) == '\n') message = message.splice(message.length-1, 1)
      logger.print('debug', 'User @' + username + ' is running command ' + activeUsers[idx].command)
      logger.print('debug', 'Sending message to child process : ' + message)

      // Send message to child process
      activeUsers[idx].child.stdin.write(message + "\n")
    }
  }

  // Is a public command?
  var isPublicCommand = function(command) {
    return contains(cm.config.commands.publicCommands, command, 'exists')
  }

  // Is a admin command?
  var isAdminCommand = function(command) {
    return contains(cm.config.commands.adminCommands, command, 'exists')
  }

  // __ Main _______________________________________________________________

  cm.addFile(configPath + 'help.json', null, true)
  cm.addFile(configPath + 'literals.json', null, true)
  cm.addFile(configPath + 'commands.json', null, true, buildCommandManager)
  var activeUsers = [] // Array of users running commands

  return { "run": run
         , "getActiveUserIdx": getActiveUserIdx
         , "removeActiveUser": removeActiveUser
         , "sendToChild": sendToChild
         , "isPublicCommand": isPublicCommand
         , "isAdminCommand": isAdminCommand
         , "publicScriptsPath": cm.config.commands.publicScriptsPath
         , "adminScriptsPath": cm.config.commands.adminScriptsPath
         }
}

module.exports = CommandManager
